pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -521,7 +521,20 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 	resources.ApplyTaskResults(nextRprts, resolvedResultRefs)
 
 	// GetFinalTasks only returns tasks when a DAG is complete
-	nextRprts = append(nextRprts, pipelineRunFacts.GetFinalTasks()...)
+	finallyRprts := pipelineRunFacts.GetFinalTasks()
+
+	// Before creating TaskRun for scheduled final task, check if it's consuming a task result
+	// Resolve and apply task result wherever applicable, report failure in case resolution fails
+	for _, rprt := range finallyRprts {
+		resolvedResultRefs, err := resources.ResolveResultRefs(pipelineRunFacts.State, resources.PipelineRunState{rprt})
+		if err != nil {
+			rprt.InvalidTaskResultsInFinally = err.Error()
+			logger.Warnf("Declaring final task %q failed as it failed to resolve task params for %q with error %v", rprt.PipelineTask.Name, pr.Name, err)
+			continue
+		}
+		resources.ApplyTaskResults(resources.PipelineRunState{rprt}, resolvedResultRefs)
```

Comment:
It feels a bit un-natural to use two functions that work on slices, passing them elements of a slice one by one embedded in 1-element slices.
I wonder if it would make sense to make a version of those functions that work on single elements, and use that both here as well as for the implementation of the current functions that work on slices.

Reply:
Not something that needs to be fixed in the PR, it's also fine to change it as a follow-up if you agree.

Reply:
I have introduced a new function to handle single task for now, `taskResults` is due for refactoring. After @sbwsg's pipelineRun results PR #3472 is merged, I will take up refactoring `taskResults`. Changes in PR #3472 have simplified resolving pipelineRun results and cleaned up `resultrefresolution.go`.